enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,369 @@
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+spec:
+  managementState: Managed
+  plugins:
+  - my-plugin
+```
+
+This will trigger a new rollout of the console `Deployment` with the updated
+plugins. Note that this does mean the user will need to refresh the browser
+to see the new plugins.
+
+An operator can also give console a hint that it has a plugin available through
+the `console.openshift.io/plugins` annotations. The annotation tells the console to
+show a checkbox to enable plugins during operator install. The value is a
+serialized JSON array of strings, where each item is the name of a
+`ConsolePlugin` resource the operator will create.
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    console.openshift.io/plugins: '["my-plugin"]'
+```
+
+For operators from the `redhat-operators` catalog source, the install operator
+page will check the plugin checkbox by default. For other operators, the
+checkbox will not be checked by default. Any plugin can be enabled or disabled
+later through the console cluster settings page.
+
+### Loading Plugins
+
+The console backend will proxy requests to the plugin `Service`. This way the
+`Service` itself does not need to be exposed outside the cluster, and avoids
+any same-origin or certificate errors in the browser loading JavaScript from a
+separate domain. 
+
+On load, console will read the manifest for each plugin. The manifest is a JSON
+document that contains metadata about the plugin and the extensions.
+`extensions` mirrors the API used for static plugins, and the same extension
+points will be available.
+
+```json
+{
+  "$schema": "/path/to/schema/manifest.json",
+  "id": "my-plugin",
+  "version": "1.2.3",
+  "pluginAPI": "4.7.x",
+  "displayName": "My Plugin",
+  "extensions": [
+    {
+      "type": "console.flag",
+      "properties": {
+        "handler": {
+          "$codeRef": "example.testHandler"
+        }
+      }
+    },
+    {
+      "type": "console.flag/model",
+      "properties": {
+        "flag": "EXAMPLE",
+        "model": {
+          "group": "example.com",
+          "version": "v1",
+          "kind": "ExampleModel"
+        }
+      }
+    }
+  ]
+}
+```
+
+#### Plugin Manifest Properties
+
+| Property | Description |
+| ----- | -------- |
+| `name: string` | Used as a unique identifier. Each plugin must have a unique name. |
+| `version: string` | The version of the plugin. Version must be parsable by node-semver, eg. “1.2.3” |
+| `pluginAPI: string` | Semver range of compatible OpenShift console versions. |
+| `displayName?: string` | User friendly display name. |
+| `description?: string` | The description of the plugin. |
+| `extensions: Extension[]` | List of extensions contributed by the plugin. |
+| `dependencies?: { [pluginName: string]: string }` | A plugin which depends on other plugins must explicitly define their dependencies. |
```

Comment:
do you have a use/need for this today?  dependency management is tricky and this looks too naive to really work (e.g. it doesn't handle version requirements) so if you don't have an explicit need for it, it might be better to punt on dep management entirely until you can come up with a more fully fleshed out design.

Reply:
I believe @christianvogt has a specific need for this in dev console.

Reply:
hm.  I still worry about how simplistic this is w/ no versioning capability.

what happens when the dependencies are not met?

Reply:
@christianvogt The version is the value in the map here correct? For instance,

```json
"dependencies": {
  "my-plugin": "1.x",
  "another-plugin": "2.3.x"
}
```

Reply:
I pushed an update improving the property description to make this clear.